cql3/statements/create_table_statement.cc

```diff
@@ -96,10 +97,31 @@ std::vector<column_definition> create_table_statement::get_columns()
     return column_defs;
 }
 
+template <typename CreateTable>
+future<shared_ptr<cql_transport::event::schema_change>>
+create_table_statement::create_table_with_cdc(service::storage_proxy& proxy,
+                                              schema_ptr schema,
+                                              CreateTable&& create_table) {
+    if (_if_not_exists) {
+        throw exceptions::invalid_request_exception(
+                "Can't create table with CDC support using IF NOT EXISTS");
+    }
+    return cdc::setup(proxy, schema).then([create_table = std::move(create_table), schema = std::move(schema)] () mutable{
+        return create_table().handle_exception([schema = std::move(schema)](std::exception_ptr ep) mutable {
+            return cdc::remove(schema->ks_name(), schema->cf_name()).then([ep = std::move(ep)] {
+                using ret_type = shared_ptr<cql_transport::event::schema_change>;
+                return make_exception_future<ret_type>(std::move(ep));
+            });
+        });
+    });
+}
+
 future<shared_ptr<cql_transport::event::schema_change>> create_table_statement::announce_migration(service::storage_proxy& proxy, bool is_local_only) {
-    auto create_table = [this, is_local_only, &proxy] {
+    auto schema = get_cf_meta_data(proxy.get_db().local());
+    auto create_table = [this, is_local_only, schema] () mutable {
```

Comment:
Why not make it a private method or a static function in this unit? What do we gain with all the lambda begin passed around? `create_table_with_cdc` can just call the method/function.

Reply:
Private method and static function work just as well. The benefit of having lambda here is encapsulation. With lambda it's blunt clear that it's just a helper method for announce_migration and nothing else. Having said that, I'm fine with private method. I can change it if you want?

Reply:
Please don't. Always keep all used entities as local as possible. If the lambda function was needed in a different function also, make it a private function. Otherwise, keep it as close to the usage as possible.

Reply:
Is there any other place in the code where we do something like that: there's a generic function (`create_table_with_cdc`) that takes a function parameter (`CreateTable&& create_table`) and is called in **exactly one** place (`announce_migration`) with **exactly one** function (the `create_table` lambda)?

It just seems like a very non-standard thing.

Also, since `create_table_with_cdc` takes a function parameter, it suggests that it's called in multiple places with different functions passed in for this parameter. And that it doesn't really care what the function parameter is.
However for `create_table_with_cdc` to be correct there is only one function that can be passed in here, which is now the `create_table` lambda. So the parametrization is artificial.

Locality can be easily achieved my using a static function. It will then be local to the compilation unit. You can easily find all usages of the function in the unit by using text search.

Reply:
Less local than inside the function. And requires search/parse of rest of compilation unit/class.